### PR TITLE
[ModelOpt] Load w13/w2_input_scale for all experts, nvfp4

### PR DESCRIPTION
**Pull request by wenscarl**
This matches trt-llm usage. Use all experts' input scales to compute alpha and quantize. This change doesn't affect perf.

## Test Plan
```
VLLM_WORKER_MULTIPROC_METHOD="spawn" \
VLLM_USE_STANDALONE_COMPILE=0 \
VLLM_USE_FLASHINFER_MOE_FP4=1 \
VLLM_FLASHINFER_MOE_BACKEND="throughput" \
lm_eval --model vllm --model_args pretrained=nvidia/DeepSeek-R1-FP4,quantization=modelopt_fp4,data_parallel_size=8,enable_expert_parallel=True,tensor_parallel_size=1,max_model_len=2048,enforce_eager=True --trust_remote_code --tasks gsm8k --num_fewshot 5 --batch_size auto
```
## Test Result
After change:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9431|±  |0.0064|
|     |       |strict-match    |     5|exact_match|↑  |0.9401|±  |0.0065|```

Before change:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.934|±  |0.0068|
|     |       |strict-match    |     5|exact_match|↑  |0.931|±  |0.0070|
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 1]
```diff
@@ -1464,7 +1466,9 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         layer.w13_weight_scale_2 = Parameter(w13_weight_scale_2, requires_grad=False)
 
         # Common processing for input scales and alphas
-        w13_input_scale = layer.w13_input_scale.max(dim=1).values.to(torch.float32)
+        w13_input_scale = (
+            layer.w13_input_scale.max().to(torch.float32).expand(layer.num_experts)
+        )
```
Comment (pavanimajety) [inline vllm/model_executor/layers/quantization/modelopt.py:1469]: Why are we performing a max and then running an expand? Also, w13_input_scale.max(dim=1) takes a max of w1 and w3 and generates [E,] where as 
layer.w13_input_scale.max() takes the global max of the stacked tensor? And additionally, why do we want to change this for all paths?

Comment (wenscarl) [reply vllm/model_executor/layers/quantization/modelopt.py:1469]: The scope should be limited to modelopt_nvfp4 quantization only.
To ensure consistency in subsequent multiplication dimensions, it’s preferable to preserve the original tensor shape as before this PR, while changing only the numerical behavior.

Comment (pavanimajety) [reply vllm/model_executor/layers/quantization/modelopt.py:1469]: My question is specific to changing how we compute the input_scale for cutlass_moe_fp4 path and the flashinfer paths. I think we need to modify only for the relevant flashinfer paths.

Comment (wenscarl) [reply vllm/model_executor/layers/quantization/modelopt.py:1469]: sure. Will make the change.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -1546,13 +1547,16 @@ def weight_loader(
...
             return True if return_success else None
 
-        expert_id = self._map_global_expert_id_to_local_expert_id(expert_id)
-        if expert_id == -1:
+        quant_method_name = self.quant_method.__class__.__name__
+        global_expert_id = expert_id
+        expert_id = self._map_global_expert_id_to_local_expert_id(global_expert_id)
+        is_modeloptnvfp4 = quant_method_name == "ModelOptNvFp4FusedMoE"
+        is_input_scale = "input_scale" in weight_name
+        if expert_id == -1 and not (is_modeloptnvfp4 and is_input_scale):
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py]: If we want to be extra paranoid, it seems like we could limit this even further by checking if `self.quant_method.allow_flashinfer and self.quant_method.flashinfer_moe_backend == FlashinferMoeBackend.CUTLASS`?

Comment (wenscarl) [reply vllm/model_executor/layers/fused_moe/layer.py]: yes. I will make it `self.quant_method.allow_flashinfer and self.quant_method.flashinfer_moe_backend in (some backends supporting  EP)` because the inconsistency of global scaling factor used in quantize and dequantize associated with EP case.

**Comment by leejnau** [comment]
For the `nvidia/DeepSeek-R1-0528-FP4-v2` model, in both TP4 and DP4 modes, with the FlashInfer backend, this PR raises the accuracy from ~2% to ~95%. 

#### server (TP4):
`VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_FLASHINFER_MOE_BACKEND="throughput" vllm serve nvidia/DeepSeek-R1-0528-FP4-v2   --quantization="modelopt_fp4"   --trust-remote-code   --gpu_memory_utilization=0.8   --tensor-parallel-size 4   --data-parallel-size 1`
#### client: 
`python3 tests/evals/gsm8k/gsm8k_eval.py`
`commit: 29350922c64a808a6de3b0e31fbadc2aebd6ba3f`
`Accuracy: 0.025`
`commit: 281be34de010fd4b106341e8aa3996f01f121c61`
`Accuracy: 0.955`

#### server (DP4):
`VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_FLASHINFER_MOE_BACKEND="throughput" vllm serve nvidia/DeepSeek-R1-0528-FP4-v2   --quantization="modelopt_fp4"   --trust-remote-code   --gpu_memory_utilization=0.8   --tensor-parallel-size 1   --data-parallel-size 4`
#### client: 
`python3 tests/evals/gsm8k/gsm8k_eval.py` 
`commit: 29350922c64a808a6de3b0e31fbadc2aebd6ba3f`
`Accuracy: 0.022`
`commit: 281be34de010fd4b106341e8aa3996f01f121c61`
`Accuracy: 0.955`

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 3]
```diff
@@ -1476,14 +1489,20 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         )
 
         # GEMM 2 processing
+        if self.allow_flashinfer:
+            w2_input_scale = (
+                layer.w2_input_scale.max().to(torch.float32).expand(layer.num_experts)
```
Comment (pavanimajety) [inline vllm/model_executor/layers/quantization/modelopt.py:1493]: Nit: Please add comment for why we are taking the max here.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 4]
```diff
@@ -1493,7 +1498,7 @@ def weight_loader(self,
 
             self._load_single_value(param=param,
                                     loaded_weight=loaded_weight,
-                                    expert_id=expert_id)
+                                    expert_id=global_expert_id if is_modeloptnvfp4 else expert_id)
             return True if return_success else None
```
[the remark this answers is not in the document]
Comment (pavanimajety) [reply vllm/model_executor/layers/fused_moe/layer.py:1746]: @leejnau / @wenscarl Could you please validate this? Ideally in post_weight_loading we want to ensure that we see different input scales for the 256x2 (w13)  expert where we take the maximum

[the remark this answers is not in the document]
Comment (wenscarl) [reply vllm/model_executor/layers/fused_moe/layer.py:1746]: Yes, in layer.py, 256 values are read in for each expert rank. In modelopt.py, max is taken among 256x2(w13).

File: vllm/model_executor/layers/fused_moe/layer.py
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:1727]: Nit: looks like these comments are out of date

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 5]
```diff
@@ -1546,13 +1550,25 @@ def weight_loader(
...
+
+        allow_flashinfer = getattr(self.quant_method, "allow_flashinfer", False)
+        moe_backend = getattr(self.quant_method, "flashinfer_moe_backend", None)
+
+        use_global_sf = (
+            allow_flashinfer
+            and is_flashinfer_supporting_global_sf(moe_backend)
+            and "input_scale" in weight_name
+            and quant_method_name == "ModelOptNvFp4FusedMoE"
+        )
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:1651]: Would be good to put these three lines together and leave a comment on what use_global_sf means in this case since we are in `fused_moe/layer.py`

**Review by mgoin — APPROVED** [review]
LGTM, although checking for local attrs in fused_moe/layer.py doesn't feel good to keep doing I don't have a better option atm
